README.md
# Yireo CorsHack
The new Magento 2.3 GraphQL system could be used with GraphQL clients
(like Apollo or even Axios) to fetch data from Magento. Most of these
clients use an HTTP request OPTIONS to see if CORS restrictions apply.
This module adds an OPTIONS check to the GraphQL API. Also, this module
adds Cross Origin headers (currently hard-coded to
`http://localhost:3000`).

### Installation
```
composer require yireo-training/magento2-corshack
./bin/magento module:enable Yireo_CorsHack
```

### Configuration
Navigate to **Advanced > Yireo CorsHack** and add your URLs to the **Origin Domain** option.

By default, the following URLs work right away:
- `http://localhost/`
- `http://localhost:3000/`

If you want all domains to match, add a wildcard (`*`) to the **Origin Domain** option.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.6] - 29 July 2020
### Added
- Magento 2.4 compatibility

## [1.0.5] - Undocumented


Test/Integration/CheckResponseTest.php
<?php
/**
 * CorsHack module for Magento 2
 *
 * @package     Yireo_CorsHack
 * @author      Yireo
 * @copyright   Copyright 2018 Yireo (https://www.yireo.com/)
 * @license     Open Source License
 */

declare(strict_types=1);

namespace Yireo\CorsHack\Test\Integration;

use Magento\Framework\App\Config\ScopeConfigInterface;
use Magento\Framework\App\Response\Http;
use Magento\TestFramework\ObjectManager;
use Magento\TestFramework\TestCase\AbstractController as ControllerTestCase;

/**
 * Class CheckResponseTest
 * @package Yireo\CorsHack\Test\Integration
 */
class CheckResponseTest extends ControllerTestCase
{
    /**
     * Test whether any response contains proper headers
     *
     * @magentoConfigFixture default/corshack/settings/origin *
     */
    public function testIfResponseContainsCrossOriginHeaders()
    {
        /** @var ScopeConfigInterface $scopeConfig */
        $scopeConfig = ObjectManager::getInstance()->get(ScopeConfigInterface::class);
        $this->assertSame('*', $scopeConfig->getValue('corshack/settings/origin'));

        $this->dispatch('/graphql');

        /** @var Http $response */
        $response = $this->getResponse();
        $this->assertSame(200, $response->getHttpResponseCode());

        $foundAccessControlAllowOrigin = false;
        $foundAccessControlAllowHeaders = false;

        /** @var \Laminas\Http\Headers $headers */
        $headers = $response->getHeaders();

        foreach ($headers as $header) {
            if ($header->getFieldName() === 'Access-Control-Allow-Origin') {
                $foundAccessControlAllowOrigin = true;
            }

            if ($header->getFieldName() === 'Access-Control-Allow-Headers') {
                $foundAccessControlAllowHeaders = true;
            }
        }

        $this->assertTrue($foundAccessControlAllowOrigin);
        $this->assertTrue($foundAccessControlAllowHeaders);
    }
}


Plugin/ControllerPlugin.php
<?php
/**
 * CorsHack module for Magento 2
 *
 * @package     Yireo_CorsHack
 * @author      Yireo
 * @copyright   Copyright 2018 Yireo (https://www.yireo.com/)
 * @license     Open Source License
 */

declare(strict_types=1);

namespace Yireo\CorsHack\Plugin;

use Magento\Framework\App\Response\Http as HttpResponse;
use Magento\Framework\App\Response\HttpInterface;
use Magento\GraphQl\Controller\GraphQl as Source;
use Yireo\CorsHack\Utils\ResponseGenerator;
use Laminas\Http\Request;

/**
 * Class ControllerPlugin
 *
 * @package Yireo\CorsHack\Plugin
 */
class ControllerPlugin
{
    /**
     * @var ResponseGenerator
     */
    private $responseGenerator;

    /**
     * @var Request
     */
    private $request;

    /**
     * ActionPlugin constructor.
     *
     * @param ResponseGenerator $responseGenerator
     * @param Request $request
     */
    public function __construct(
        ResponseGenerator $responseGenerator,
        Request $request
    ) {
        $this->responseGenerator = $responseGenerator;
        $this->request = $request;
    }

    /**
     * @param Source $source
     * @param HttpResponse $response
     * @return HttpInterface
     */
    public function afterDispatch(Source $source, HttpResponse $response)
    {
        if ($this->spoofStatusHeaderIgnorantly($response)) {
            $response->setStatusHeader(200);
        }

        return $this->responseGenerator->modifyResponse($response);
    }

    /**
     * @param HttpResponse $response
     * @return bool
     */
    private function spoofStatusHeaderIgnorantly(HttpResponse $response): bool
    {
        if ($this->request->getMethod() === 'OPTIONS') {
            return true;
        }

        if ($this->request->getMethod() === 'GET') {
            return true;
        }

        $data = json_decode($response->getBody(), true);
        if (!empty($data) && !empty($data['data']) && empty($data['errors'])) {
            return true;
        }

        return false;
    }
}


Utils/ResponseGenerator.php
<?php
declare(strict_types=1);

namespace Yireo\CorsHack\Utils;

use Magento\Framework\App\Config\ScopeConfigInterface;
use Magento\Framework\App\Response\HttpInterface as HttpResponse;

/**
 * Class ResponseGenerator
 * @package Yireo\CorsHack\Utils
 */
class ResponseGenerator
{
    /**
     * @var ScopeConfigInterface
     */
    private $scopeConfig;

    /**
     * HeaderGenerator constructor.
     * @param ScopeConfigInterface $scopeConfig
     */
    public function __construct(
        ScopeConfigInterface $scopeConfig
    ) {
        $this->scopeConfig = $scopeConfig;
    }

    /**
     * @param HttpResponse $response
     * @return HttpResponse
     */
    public function modifyResponse(HttpResponse $response): HttpResponse
    {
        $domains = $this->getAccessControlAllowOriginDomains();
        $response->setHeader('Access-Control-Allow-Origin', implode(', ', $domains));

        $headers = $this->getAccessControlAllowHeaders();
        $response->setHeader('Access-Control-Allow-Headers', implode(',', $headers), true);
        $response->setHeader('Access-Control-Allow-Credentials', 'true');

        return $response;
    }

    /**
     * @return string
     */
    private function getAccessControlAllowOriginDomains(): array
    {
        $domains = [];

        $storedOrigins = (string) $this->scopeConfig->getValue('corshack/settings/origin');
        $storedOrigins = explode(',', $storedOrigins);
        foreach ($storedOrigins as $storedOrigin) {
            $storedOrigin = trim($storedOrigin);
            if (!empty($storedOrigin)) {
                $domains[] = $storedOrigin;
            }
        }

        $domains = array_unique($domains);

        if (count($domains) > 1) {
            $domains = ['*'];
        }

        return $domains;
    }

    /**
     * @return array
     */
    private function getAccessControlAllowHeaders(): array
    {
        $headers = [];
        $headers[] = 'Content-Type';
        $headers[] = 'Authorization';

        return $headers;
    }
}


Plugin/GraphQLController.php
<?php
namespace Yireo\CorsHack\Plugin;

use Magento\Framework\App\Request\Http;
use Magento\Framework\App\RequestInterface;
use Magento\Framework\App\ResponseInterface;
use Magento\GraphQl\Controller\GraphQl as Source;
use Magento\Framework\Webapi\Response;

/**
 * Class GraphQLController
 * @package Yireo\CorsHack\Plugin
 */
class GraphQLController
{
    /**
     * @var ResponseInterface
     */
    private $response;

    /**
     * GraphQLController constructor.
     * @param Response $response
     */
    public function __construct(
        Response $response
    ) {
        $this->response = $response;
    }

    /**
     * @param Source $source
     * @param $closure
     * @param RequestInterface $request
     * @return Response
     */
    public function aroundDispatch(Source $source, $closure, RequestInterface $request)
    {
        /** @var $request Http */
        if ($request->isOptions()) {
            $this->response->setHttpResponseCode(200);
            return $this->response;
        }

        /** @var Response $response */
        $response = $closure($request);
        $response->setHeader('Access-Control-Allow-Origin', $this->getAccessControlAllowOrigin(), true);
        $response->setHeader('Access-Control-Allow-Headers', $this->getAccessControlAllowHeaders(), true);

        return $response;
    }

    /**
     * @todo Make a configuration option for this
     * @return string
     */
    private function getAccessControlAllowOrigin()
    {
        return 'http://localhost:3000';
    }

    /**
     * @todo Make a configuration option for this
     * @return string
     */
    private function getAccessControlAllowHeaders()
    {
        return 'Content-Type';
    }
}

Plugin/ActionPlugin.php
<?php
/**
 * CorsHack module for Magento 2
 *
 * @package     Yireo_CorsHack
 * @author      Yireo
 * @copyright   Copyright 2018 Yireo (https://www.yireo.com/)
 * @license     Open Source License
 */

declare(strict_types=1);

namespace Yireo\CorsHack\Plugin;

use Magento\Framework\App\Config\ScopeConfigInterface;
use Magento\Framework\App\Request\Http;
use Magento\Framework\App\RequestInterface;
use Magento\Framework\App\Response\Http as HttpResponse;
use Magento\Framework\App\ResponseInterface;
use Magento\Framework\App\Action\Action as Source;
use Magento\Framework\App\Response\Http as Response;

/**
 * Class ActionPlugin
 *
 * @package Yireo\CorsHack\Plugin
 */
class ActionPlugin
{
    /**
     * @var ResponseInterface
     */
    private $response;

    /**
     * @var ScopeConfigInterface
     */
    private $scopeConfig;

    /**
     * ActionPlugin constructor.
     *
     * @param Response $response
     * @param ScopeConfigInterface $scopeConfig
     */
    public function __construct(
        Response $response,
        ScopeConfigInterface $scopeConfig
    ) {
        $this->response = $response;
        $this->scopeConfig = $scopeConfig;
    }

    /**
     * @param Source $source
     * @param $closure
     * @param RequestInterface $request
     *
     * @return HttpResponse
     */
    public function aroundDispatch(Source $source, $closure, RequestInterface $request)
    {
        /** @var $request Http */
        if ($request->isOptions()) {
            $this->response->setHttpResponseCode(200);
            return $this->response;
        }

        /** @var HttpResponse $response */
        $response = $closure($request);
        $response->setHeader('Access-Control-Allow-Origin', $this->getAccessControlAllowOrigin(), true);
        $response->setHeader('Access-Control-Allow-Headers', $this->getAccessControlAllowHeaders(), true);

        return $response;
    }

    /**
     * @return string
     */
    private function getAccessControlAllowOrigin(): string
    {
        $allowOrigin = [];
        $allowOrigin[] = 'http://localhost';
        $allowOrigin[] = 'http://localhost:3000';

        $storedOrigins = (string) $this->scopeConfig->getValue('corshack/settings/origin');
        $storedOrigins = explode(',', $storedOrigins);
        foreach ($storedOrigins as $storedOrigin) {
            $storedOrigin = trim($storedOrigin);
            if (!empty($storedOrigin)) {
                $allowOrigin[] = $storedOrigin;
            }
        }

        $allowOrigin = array_unique($allowOrigin);

        // If the wildcard is here, we can remove all other URLs
        if (in_array('*', $allowOrigin)) {
            $allowOrigin = ['*'];
        }

        return implode(', ', $allowOrigin);
    }

    /**
     * @return string
     */
    private function getAccessControlAllowHeaders(): string
    {
        return 'Content-Type';
    }
}
